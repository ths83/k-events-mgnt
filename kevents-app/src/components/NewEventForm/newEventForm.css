.input {
  display: grid;
  margin-bottom: 1rem;
}

.buttons {
  margin-top: 3rem;
  display: flex;
  justify-content: center;
}

.notification {
  margin-top: 1rem;
  display: flex;
  justify-content: center;
}

#reset {
  margin-right: 2rem;
}

#submit {
  margin-left: 2rem;
}
