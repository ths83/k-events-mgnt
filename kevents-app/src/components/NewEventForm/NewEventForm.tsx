import { Button, DatePicker } from "antd";
import Title from "antd/lib/typography/Title";
import axios from "axios";
import { ErrorMessage, Field, Form, Formik } from "formik";
import { isEmpty, noop } from "lodash";
import moment from "moment";
import React, { useState } from "react";
import * as Yup from "yup";
import { datePattern } from "../../utils/date/dateHelper";
import CustomErrorMessage from "../errorMessage/CustomErrorMessage";
import "./newEventForm.css";

const NewEventForm = () => {
  const requiredFieldErrorMsg = "This field is required";
  const nameErrorMsg = "Must be 32 characters or less";

  const initialValues = {
    name: "",
    description: "",
    startDate: "",
    endDate: "",
  };

  const formValidationSchema = Yup.object().shape({
    name: Yup.string()
      .min(1, nameErrorMsg)
      .max(32, nameErrorMsg)
      .required(requiredFieldErrorMsg)
      .trim(),
    description: Yup.string().required(requiredFieldErrorMsg).trim(),
    startDate: Yup.string().required(requiredFieldErrorMsg).trim(),
    endDate: Yup.string().required(requiredFieldErrorMsg).trim(),
  });

  const compareDates = (start: string | null, end: string | null) => {
    const isValid =
      isEmpty(start) || isEmpty(end) ? false : moment(start) < moment(end);
    setDateError(isValid);
  };

  const [dateError, setDateError] = useState(false);
  const [networkError, setNetworkError] = useState("");
  const [submitting, setSubmitting] = useState(false);
  const [success, setSuccess] = useState(false);

  const containsNetworkError = !isEmpty(networkError);
  const hasErrors = dateError || containsNetworkError;

  return (
    <>
      <Title level={3}>New event form</Title>
      <Formik
        initialValues={initialValues}
        validationSchema={formValidationSchema}
        onSubmit={noop}
      >
        {({
          setErrors,
          setTouched,
          setValues,
          values,
          errors,
          setFieldValue,
        }) => (
          <Form>
            <div className="input">
              <Title level={5} id="eventNameTitle">
                Name
              </Title>
              <Field name="name" placeholder="Enter an event name" />
              <ErrorMessage name="name" />
            </div>

            <div className="input">
              <Title level={5} id="eventDescTitle">
                Description
              </Title>
              <Field
                name="description"
                placeholder="Enter an event description"
              />
              <ErrorMessage name="description" />
            </div>

            <div className="input">
              <Title level={5} id="eventStartDateTitle">
                Start Date
              </Title>
              <DatePicker
                id="startDate"
                showTime
                format={datePattern}
                onChange={(value) => {
                  setFieldValue("startDate", value?.toISOString());
                  compareDates(values.startDate, values.endDate);
                }}
                onOk={(value) => {
                  setFieldValue("startDate", value.toISOString());
                  compareDates(values.startDate, values.endDate);
                }}
              />
              <ErrorMessage name="startDate" />
            </div>

            <div className="input">
              <Title level={5} id="eventEndDateTitle">
                End Date
              </Title>
              <DatePicker
                id="endDate"
                showTime
                format={datePattern}
                onChange={(value) => {
                  setFieldValue("endDate", value?.toISOString());
                  compareDates(values.startDate, values.endDate);
                }}
                onOk={(value) => {
                  setFieldValue("endDate", value.toISOString());
                  compareDates(values.startDate, values.endDate);
                }}
              />
              <ErrorMessage name="endDate" />
            </div>

            <div className="buttons">
              <Button
                id="reset"
                shape="round"
                type="primary"
                danger
                onClick={() => {
                  setValues(initialValues);
                  setErrors({});
                  setTouched({});
                  setDateError(false);
                  setNetworkError("");
                  setSuccess(false);
                }}
              >
                Reset
              </Button>
              <Button
                id="submit"
                shape="round"
                type="primary"
                loading={submitting}
                disabled={
                  !isEmpty(errors) ||
                  isEmpty(values.description) ||
                  isEmpty(values.name) ||
                  isEmpty(values.startDate) ||
                  isEmpty(values.endDate) ||
                  dateError
                }
                onClick={() => {
                  setSubmitting(true);
                  axios
                    .post(`${process.env.REACT_APP_API_URL}/events`, {
                      name: values.name,
                      description: values.description,
                      startDate: values.startDate,
                      endDate: values.endDate,
                    })
                    .then(() => {
                      setValues(initialValues);
                      setErrors({});
                      setTouched({});
                      setSuccess(true);
                    })
                    .catch((error) => {
                      setNetworkError(error.response.data.message);
                    })
                    .finally(() => setSubmitting(false));
                }}
              >
                Submit
              </Button>
            </div>
            {hasErrors && (
              <CustomErrorMessage
                message={
                  dateError
                    ? "End date must be more recent than start date!"
                    : networkError
                }
              />
            )}
            {success && (
              <Title
                level={5}
                id="successMessage"
                className="notification"
                type="success"
              >
                Successfully created new event!
              </Title>
            )}
          </Form>
        )}
      </Formik>
    </>
  );
};

export default NewEventForm;
