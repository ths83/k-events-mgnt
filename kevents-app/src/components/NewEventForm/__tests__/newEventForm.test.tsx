import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import NewEventForm from "../NewEventForm";
import renderer from "react-test-renderer";

const expectedTitle = "New event form";
const expectedNameInputTitle = "Name";
const expectedDescInputTitle = "Description";
const expectedStartDateInputTitle = "Start Date";
const expectedEndDateInputTitle = "End Date";
const expectedResetButton = "Reset";
const expectedSubmitButton = "Submit";

const expectedNameInputPlaceholder = "Enter an event name";
const expectedDescInputPlaceholder = "Enter an event description";
const expectedDateInputPlaceholder = "Select date";

const expectedRequiredFieldMsg = "This field is required";

test("renders component", () => {
  const item = renderer.create(<NewEventForm />).toJSON();
  expect(item).toMatchSnapshot();
});

test("renders form", () => {
  render(<NewEventForm />);
  const title = screen.getByText(expectedTitle);
  const name = screen.getByText(expectedNameInputTitle);
  const namePlaceholder = screen.getByPlaceholderText(
    expectedNameInputPlaceholder
  );
  const desc = screen.getByText(expectedDescInputTitle);
  const descPlaceholder = screen.getByPlaceholderText(
    expectedDescInputPlaceholder
  );
  const startDate = screen.getByText(expectedStartDateInputTitle);
  const endDate = screen.getByText(expectedEndDateInputTitle);
  const resetButton = screen.getByText(expectedResetButton);
  const submitButton = screen.getByText(expectedSubmitButton);

  const datePlaceholder = screen.getAllByPlaceholderText(
    expectedDateInputPlaceholder
  );

  expect(title).toBeInTheDocument();
  expect(name).toBeInTheDocument();
  expect(namePlaceholder).toBeInTheDocument();
  expect(desc).toBeInTheDocument();
  expect(descPlaceholder).toBeInTheDocument();
  expect(startDate).toBeInTheDocument();
  expect(endDate).toBeInTheDocument();
  expect(resetButton).toBeInTheDocument();
  expect(submitButton).toBeInTheDocument();
  expect(datePlaceholder.length).toBe(2);
});

describe("Input", () => {
  test("displays an error message when the description field is touched only", async () => {
    render(<NewEventForm />);
    const descPlaceholder = screen.getByPlaceholderText(
      expectedDescInputPlaceholder
    );
    fireEvent.blur(descPlaceholder);

    expect(await screen.findByText(expectedRequiredFieldMsg)).not.toBeNull();
  });

  test("displays an error message when the name field is touched only", async () => {
    render(<NewEventForm />);
    const namePlaceholder = screen.getByPlaceholderText(
      expectedNameInputPlaceholder
    );
    fireEvent.blur(namePlaceholder);

    expect(await screen.findByText(expectedRequiredFieldMsg)).not.toBeNull();
  });
});
