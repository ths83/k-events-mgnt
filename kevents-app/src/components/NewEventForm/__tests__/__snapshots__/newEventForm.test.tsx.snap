// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
Array [
  <h3
    className="ant-typography"
    onClick={[Function]}
    style={
      Object {
        "WebkitLineClamp": undefined,
      }
    }
  >
    New event form
  </h3>,
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="input"
    >
      <h5
        className="ant-typography"
        id="eventNameTitle"
        onClick={[Function]}
        style={
          Object {
            "WebkitLineClamp": undefined,
          }
        }
      >
        Name
      </h5>
      <input
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter an event name"
        value=""
      />
    </div>
    <div
      className="input"
    >
      <h5
        className="ant-typography"
        id="eventDescTitle"
        onClick={[Function]}
        style={
          Object {
            "WebkitLineClamp": undefined,
          }
        }
      >
        Description
      </h5>
      <input
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter an event description"
        value=""
      />
    </div>
    <div
      className="input"
    >
      <h5
        className="ant-typography"
        id="eventStartDateTitle"
        onClick={[Function]}
        style={
          Object {
            "WebkitLineClamp": undefined,
          }
        }
      >
        Start Date
      </h5>
      <div
        className="ant-picker"
        onMouseUp={[Function]}
      >
        <div
          className="ant-picker-input"
        >
          <input
            autoComplete="off"
            id="startDate"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            placeholder="Select date"
            readOnly={true}
            size={18}
            title=""
            value=""
          />
          <span
            className="ant-picker-suffix"
          >
            <span
              aria-label="calendar"
              className="anticon anticon-calendar"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="calendar"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      className="input"
    >
      <h5
        className="ant-typography"
        id="eventEndDateTitle"
        onClick={[Function]}
        style={
          Object {
            "WebkitLineClamp": undefined,
          }
        }
      >
        End Date
      </h5>
      <div
        className="ant-picker"
        onMouseUp={[Function]}
      >
        <div
          className="ant-picker-input"
        >
          <input
            autoComplete="off"
            id="endDate"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            placeholder="Select date"
            readOnly={true}
            size={18}
            title=""
            value=""
          />
          <span
            className="ant-picker-suffix"
          >
            <span
              aria-label="calendar"
              className="anticon anticon-calendar"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="calendar"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      className="buttons"
    >
      <button
        className="ant-btn ant-btn-primary ant-btn-round ant-btn-dangerous"
        id="reset"
        onClick={[Function]}
        type="button"
      >
        <span>
          Reset
        </span>
      </button>
      <button
        className="ant-btn ant-btn-primary ant-btn-round"
        disabled={true}
        id="submit"
        onClick={[Function]}
        type="button"
      >
        <span>
          Submit
        </span>
      </button>
    </div>
  </form>,
]
`;
